### ISSUE TITLE
`PooledHttpFile` and `PooledAggregatedHttpFile` ?

### ISSUE THREAD

**Issue by trustin** [line/armeria#2863]
While working on some fixes related with `HttpFile`, I realized that we have `aggregateWithPooledObjects()` in `HttpFile`. Does this mean we need to introduce `Pooled*` variants for `HttpFile` and `AggregatedHttpFile` like we did for `HttpRequest` and `AggregatedHttpRequest`?

One core difference between `HttpRequest` and `HttpFile` is that `HttpFile.aggregate()` implementation is currently tightly coupled with `HttpFile` itself, e.g. `StreamingHttpFile` implements `aggregate(WithPooledObjects)` by itself, not using a generic aggregator implementation. As a result, it is not possible to provide a generic `Pooled(Aggregated)HttpFile` implementation like we did for `(Aggregated)HttpRequest`.

There are a few solutions for this issue in my opinion:

- Assume all `(Aggregated)HttpFile` implementations also implement `Pooled(Aggregated)HttpFile`, so that `Pooled(Aggregated)HttpFile.of()` always just downcasts. I'm not sure I like this idea because we enforce implementation detail to those who implement `HttpFile`.
- Add an operation to `HttpFile`, that allows us to write some generic aggregation logic like we did for `HttpRequest`. However, we lose an opportunity to optimize the aggregation logic. For example, when we can generate an `AggregatedHttpFile` from `HttpDataFile` really trivially.
  - I see the same issue in `HttpRequest.aggregate()`, e.g. we could do a lot more optimization for  `OneElementFixedHttpRequest`.
- Add `toPooled()` method and remove the static factory methods. However, we ditched this idea before.

Thoughts? /cc @line/dx @anuraaga

**Comment by minwoox** [comment]
> Add toPooled() method and remove the static factory methods. However, we ditched this idea before.

Yes, we have ditched the idea and I agreed on it.
However, reference counting is getting well known nowadays, and I think it's not so much bad adding `toPooled()` method if we put up a warning message on Javadoc.

> we lose an opportunity to optimize the aggregation logic

I don't want to lose that opportunity and I want to optimization for `OneElementFixedHttpRequest`. 😉

**Comment by anuraaga** [comment]
> However, reference counting is getting well known nowadays

Don't agree with this at all - this is Java, reference counting is not idiomatic in any way and never will be ;)

That doesn't mean having a strong warning in javadoc isn't enough to discourage it's use, but how is `toPooled` different from a factory in this case? Are there other options that can still keep the separation better, like an `InternalFoo` class (I think we've avoided these, but maybe there are some good uses for it)?

**Comment by minwoox** [comment]
> but how is toPooled different from a factory in this case?

If so, the class implementation can use internal fields to implement the method. That's what I figured.

> this is Java, reference counting is not idiomatic in any way and never will be ;)

Yes, of course. :)

> That doesn't mean having a strong warning in javadoc isn't enough to discourage it's use,

But I think we're already doing that in `RequestContext` by providing `alloc()` method with the strong warning. Of course, a user has to use the allocator as a parameter of another method such as `aggregateWithPooledObjects`, but it's still true that we don't hide it from the public APIs.

**Comment by trustin** [comment]
> how is toPooled different from a factory in this case? Are there other options that can still keep the separation better, like an `InternalFoo` class (I think we've avoided these, but maybe there are some good uses for it)?

I guess you might have considered this, but just in case... An implementation can override `toPooled()` to yield its `Pooled*` implementation that is optimized or that needs to access the internal details of the implementation.

In the case of `HttpFile`, it is currently impossible to implement `aggregateWithPooledObject()` without accessing the internal implementation. We'll need to refactor a lot, like generalizing `HttpFile` into something like `StreamMessage<HttpFileObject>` to achieve that, but I'm not sure if it's worth doing so.

Even if we have done such refactoring, a general aggregator implementation will always worse than implementation-specific `aggregate*()` implementation, like it is currently in `PooledHttp{Request,Response}.aggregate*()`.

**Comment by anuraaga** [comment]
> needs to access the internal details of the implementation

Yeah was thinking it should be possible to do this but, for example, make it package private and only accessible through an `InternalAccess` type of class in the same package that is for internal use only so at least there's separation by class. I don't think we've used the pattern so far but an option to consider. Maybe not worth it.

I feel like the advantage of separation in #2448 goes away with `toPooled` so if that were to happen, could just revert it to fix the issue about wrapping object allocations at the same time?

**Comment by minwoox** [comment]
> could just revert it to fix the issue about wrapping object allocations at the same time?

Oh, please don't say that. 🤣
Well, I didn't get quite understanding of `InterfalFoo`, could you elaborate more on it, please?

**Comment by anuraaga** [comment]
> Or an example?

Yeah, I think it's something like this

```java
package com.linecorp.armeria.common;

class OneElementFixedHttpRequest {
  // package private
  PooledHttpRequest toPooled() {
    return ...
  }
}

// For internal use only
public class InternalAccess {
  public PooledHttpRequest toPooled(HttpRequest req) {
    if (req instanceof OneElementFixedHttpRequest) {
      return (..)req.toPooled();
    }
  }
}

package com.linecorp.armeria.common.unsafe;

public class PooledHttpRequest {
  public static PooledHttpRequest of(HttpRequest req) { 
     InternalAccess.toPooled(req);
  }
}
```

It's definitely some more boilerplate, but is a pattern that can keep the `toPooled` type of logic to the `unsafe` package and this `Internal` class

**Comment by minwoox** [comment]
Oh, thanks!
I think it's worth trying and it might solve the problem we are facing here. 👍

**Comment by trustin** [comment]
IIUC, we cannot provide any optimization for user-implemented ones if we do that. Are we sure we want that?

**Comment by anuraaga** [comment]
Actually, that brings a different idea that might be less boilerplate too and should support them. Can we add `Poolable` interface in the `unsafe` package with `toPooled` and check `instanceof Poolable` before wrapping with the default?

**Comment by trustin** [comment]
While trying the idea of `Poolable`, I realized that it won't work when a user is using `PublishedBasedStreamMessage`. :sob:

**Comment by trustin** [comment]
Probably solved by #2882 and #2892. Let's continue the discussion at #2892.
